### PR TITLE
Add an executable specifically for ubuntu 20.04 and 22.04

### PR DESCRIPTION
**Pull request by valyo**
## **1. This PR contains the following changes...**

User reported that the latest ubuntu executable generates a python-related error on Ubuntu 20.04 (https://github.com/ScilifelabDataCentre/dds_cli/issues/600). This PR provides a solution in the form of publishing a separate executables for Ubuntu latest (22.04 currently) and Ubuntu 20.04

## **3. The PR fixes the following GitHub issue / Jira task**

- [x] GitHub issue (link): https://github.com/ScilifelabDataCentre/dds_cli/issues/600
- [x] Jira task (ID, DDS-1463):
- [ ] The PR does not fix a specific GitHub issue or Jira task

## **4. What _type of change(s)_ does the PR contain?**

- [ ] New feature
  - [ ] Breaking: _Please describe the reason for the break and how we can fix it._
  - [ ] Non-breaking
- [ ] Bug fix
  - [ ] Breaking: _Please describe the reason for the break and how we can fix it._
  - [ ] Non-breaking
- [ ] Security Alert fix
- [ ] Documentation
- [ ] Tests **(only)**
- [x] Workflow

### **Always**

- [Changelog](../CHANGELOG.md)
  - [x] Added
  - [ ] Not needed (E.g. PR contains _only_ tests)
- Rebase / Update / Merge _from_ base branch (the branch from which the current is forked)
  - [x] Done
  - [ ] Not needed
- Blocking PRs
  - [x] Merged
  - [ ] No blocking PRs
- PR to `master` branch
  - [ ] Yes: Go to the section [PR to master](#pr-to-master)
  - [x] No

### If PR consists of **code change(s)**

- Self review
  - [x] Done
- Comments, docstrings, etc
  - [x] Added / Updated
- Documentation
  - [x] Updated
  - [ ] Update not needed

### If PR is to **master**

- [ ] I have followed steps 1-5 in [the release instructions](../docs/procedures/new_release.md)
- [ ] I am bumping the major version (e.g. 1.x.x to 2.x.x)
- [ ] I have made the corresponding changes to the web/API version

**Is this version _backward compatible?_**

- [ ] Yes: The code works together with `dds_web/master` branch
- [ ] No: The code **does not** entirely / at all work together with the `dds_web/master` branch. _Please add detailed and clear information about the broken features_

## **6. Actions / Scans**

- **Black**: Python code formatter. Does not execute. Only tests.
  Run `black .` locally to execute formatting.
  - [x] Passed
- **Prettier**: General code formatter. Our use case: MD and yaml mainly.
  Run `npx prettier --write .` locally to execute formatting.
  - [x] Passed
- **Tests**: Pytest to verify that functionality works as expected.
  - [ ] New tests added
  - [x] No new tests
  - [ ] Passed
- **TestPyPi**: Build CLI and publish to TestPyPi in order to verify before release.
  - [x] Passed
- **CodeQL**: Scan for security vulnerabilities, bugs, errors
  - [ ] New alerts: _Go through them and either fix, dismiss och ignore. Add reasoning in items below._
  - [ ] Alerts fixed: _What?_
  - [ ] Alerts ignored / dismissed: _Why?_
  - [ ] Passed
- **Trivy**: Security scanner
  - [ ] New alerts: _Go through them and either fix, dismiss och ignore. Add reasoning in items below._
  - [ ] Alerts fixed: _What?_
  - [ ] Alerts ignored / dismissed: _Why?_
  - [ ] Passed
- **Snyk**: Security scanner
  - [ ] New alerts: _Go through them and either fix, dismiss och ignore. Add reasoning in items below._
  - [ ] Alerts fixed: _What?_
  - [ ] Alerts ignored / dismissed: _Why?_
  - [ ] Passed

### CODE REVIEW

File: .github/workflows/release-cli.yml [hunk 1]
```diff
@@ -70,7 +70,7 @@ jobs:
             CMD_BUILD: PyInstaller -F -c -n dds_cli_win_x86_64.exe -i resources/scilifelab.ico  --exclude-module=tests --log-level INFO dds_cli/__init__.py
             OUT_FILE_NAME: dds_cli_win_x86_64.exe
             ASSET_MIME: application/vnd.microsoft.portable-executable
-          - os: ubuntu-latest
+          - os: ubuntu-22.04, ubuntu-20.04
```
Comment (i-oden) [inline .github/workflows/release-cli.yml:73]: I think we need one for each right? Otherwise both will have the same OUT_FILE_NAME below

Comment (valyo) [reply .github/workflows/release-cli.yml:73]: you mean one separate `- os:` block for each, yes that makes sense

**Comment by valyo** [comment]
Is there a good way to upload the assets without making a new release?
This is confusing me a bit since it is probably not a good idea to add assets on PR to the latest release, but can we have somewhere else?

File: .github/workflows/release-cli.yml (hunk 1 above)
Comment (i-oden) [reply .github/workflows/release-cli.yml:73]: Yeah, thought of that too. Go with `ubuntu-latest` and `ubuntu-20.04`

Comment (i-oden) [reply .github/workflows/release-cli.yml:73]: And add a comment by `ubuntu-latest` that it's `22.04` at the moment of writing

**Comment by i-oden** [comment]
> Is there a good way to upload the assets without making a new release? This is confusing me a bit since it is probably not a good idea to add assets on PR to the latest release, but can we have somewhere else?

I'm not entirely sure. Try copying the `ddsclibinaries:` part to a new file, running it locally (should be possible right?) without uploading it anywhere and if that works and generates the additional executable for 20.04 (and it works when running) then we just update the `release-cli.yml` and then make a new minor release.

**Comment by valyo** [comment]
> Try copying the `ddsclibinaries:` part to a new file, running it locally (should be possible right?)

not sure if/how this can be run locally, though

**Comment by i-oden** [comment]
> > Try copying the `ddsclibinaries:` part to a new file, running it locally (should be possible right?)
> 
> not sure if/how this can be run locally, though I mean, using just a github workflow .yml file The only way I can imagine is setting up a docker container and doing it

True. Better way is to run the pyinstaller command locally instead (CMD_BUILD)

**Comment by i-oden** [comment]
@valyo So running 
```
pyinstaller -F -c -n  dds_cli_ubuntu_x86_64 -i resources/scilifelab.icns  --exclude-module=tests  --target-arch x86_64 --log-level INFO dds_cli/__init__.py
```
locally on 20.04 should create an executable working for it, or there should be a pyinstaller setting that lets you specify the os.

**Comment by i-oden** [comment]
> yes, but locally for me is mac os arm architecture, and if I understand it correctly it shouldn't be possible to build for ubuntu 20.04 also, having it mind what _nylander_ wrote in the issue, I doubt it that building in this way will prove that we will have functioning binaries resulting from our github action maybe we just to make a release, and it should work

Yeah, I meant that you'd run it where ever you reproduced the error before. 
I mean since it's the exact same command that the action runs it should prove that it'd work from the action. We can also make a fork of the repo to our own profile, comment out everything regarding documentation and publishing to PyPI / TestPyPI and then make a release in your fork. I don't really the idea of making a new release without testing it first, if it for some reason doesn't work we'll be making multiple new versions.

**Comment by i-oden** [comment]
```yaml
name: Make executables in fork
on:
  release:
    types: [published]
    branches:
      - "master"
jobs:
  ddsclibinaries:
    name: Build binary packages for the DDS CLI
    needs: build-n-publish
    runs-on: ${{ matrix.os }}
    strategy:
      matrix:
        include:
          - os: macos-latest
            TARGET: MacOs_x86_64
            CMD_BUILD: pyinstaller -F -c -n  dds_cli_macos_x86_64 -i resources/scilifelab.icns  --exclude-module=tests  --target-arch x86_64 --log-level INFO dds_cli/__init__.py
            OUT_FILE_NAME: dds_cli_macos_x86_64
            ASSET_MIME: application/x-elf
          #- os: macos-latest  #No virtual environment for this yet
          #  TARGET: MacOs_arm64
          #  CMD_BUILD: pyinstaller -F -c -n dds_cli_macos_arm64 -i resources/scilifelab.icns --exclude-module=tests --target-arch arm64 --log-level INFO dds_cli/__init__.py
          #  OUT_FILE_NAME: dds_cli_macos_arm64
          #  ASSET_MIME: application/x-elf
          - os: windows-latest
            TARGET: Windows_x86_64
            CMD_BUILD: PyInstaller -F -c -n dds_cli_win_x86_64.exe -i resources/scilifelab.ico  --exclude-module=tests --log-level INFO dds_cli/__init__.py
            OUT_FILE_NAME: dds_cli_win_x86_64.exe
            ASSET_MIME: application/vnd.microsoft.portable-executable
          - os: ubuntu-20.04
            TARGET: Linux_x86_64
            CMD_BUILD: pyinstaller -F -c -n  dds_cli_ubuntu_x86_64 -i resources/scilifelab.icns  --exclude-module=tests  --target-arch x86_64 --log-level INFO dds_cli/__init__.py
            OUT_FILE_NAME: dds_cli_ubuntu-20.04_x86_64
            ASSET_MIME: application/x-elf
          - os: ubuntu-latest # ubuntu-22.04 at the moment
            TARGET: Linux_x86_64
            CMD_BUILD: pyinstaller -F -c -n  dds_cli_ubuntu_x86_64 -i resources/scilifelab.icns  --exclude-module=tests  --target-arch x86_64 --log-level INFO dds_cli/__init__.py
            OUT_FILE_NAME: dds_cli_ubuntu-latest_x86_64
            ASSET_MIME: application/x-elf
    steps:
      - uses: actions/checkout@v2
      - name: Python 3.9 setup
        uses: actions/setup-python@v2
        with:
          python-version: 3.9
          # architecture: 'x64'
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install PyInstaller
          pip install -r requirements.txt
      - name: Building with PyInstaller for ${{matrix.TARGET}}
        run: ${{matrix.CMD_BUILD}}
      - name: Test the executable
        run: |
          ./dist/${{ matrix.OUT_FILE_NAME}} --version
      - name: Upload Release Asset
        id: upload-release-asset
        uses: actions/upload-release-asset@v1
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        with:
          upload_url: ${{ github.event.release.upload_url }}
          asset_path: ./dist/${{ matrix.OUT_FILE_NAME}}
          asset_name: ${{ matrix.OUT_FILE_NAME}}
          asset_content_type: ${{ matrix.ASSET_MIME}}
```

File: .github/workflows/release-cli.yml [hunk 2]
```diff
@@ -77,12 +77,19 @@ jobs:
               --exclude-module=tests --log-level INFO dds_cli/__init__.py
             OUT_FILE_NAME: dds_cli_win_x86_64.exe
             ASSET_MIME: application/vnd.microsoft.portable-executable
-          - os: ubuntu-latest
+          - os: ubuntu-20.04
             TARGET: Linux_x86_64
             CMD_BUILD: >
-              pyinstaller -F -c -n  dds_cli_ubuntu_x86_64 -i resources/scilifelab.icns
+              PyInstaller -F -c -n  dds_cli_ubuntu-20.04_x86_64 -i resources/scilifelab.icns
```
Comment (i-oden) [inline .github/workflows/release-cli.yml:83]: Switch to `pyinstaller` here or to `PyInstaller` on the other rows (assuming it doesn't matter and still works)

**Review by i-oden — CHANGES_REQUESTED** [review]
@valyo Just one minor change, but looks good!

**Review by i-oden — CHANGES_REQUESTED** [review]
Just add the changelog row and I'm ready to approve and merge

**Review by i-oden — APPROVED** [review]
Looks good!

**Pull request merged**